### ISSUE TITLE
Dependency Injection Documentation

### ISSUE THREAD

**Issue by dschreck17** [angular/angular#19749]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

Documentation in question: https://angular.io/guide/dependency-injection-in-action
"Now that you've registered these services, Angular can inject them into the constructor of any component or service, anywhere in the application." You still have to import the service in every component you wish to use it.

## Expected behavior

I expect to import the service in the root app component and not have to call "import servicename from dir in every component I wish to use it. 

By including the service in the constructor,  I should be able to use the service in this component without any additional import statements.

## Minimal reproduction of the problem with instructions

look at active demo here: https://angular.io/generated/zips/dependency-injection-in-action/dependency-injection-in-action.zip and look at src/app/hero-bios.component.ts you still have to import UserService. 

## What is the motivation / use case for changing the behavior?

What's the point of importing the service into the main app component if I have to call "import service from dir" in every component. I wish I could just import the service in my root component and then include the service in any component constructor and not have to "import"

## Environment

<pre><code>
Angular version: 4.4.5
Angular CLI: 1.4.5

Browser:
- [x] Chrome (desktop) version 59.0.3071.86
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 7.40  
- Platform: Mac 

Others:

</code></pre>

**Comment by Toxicable** [comment]
>I expect to import the service in the root app component and not have to call "import servicename from dir in every component I wish to use it.

Sorry but this is never going to happen.
This isn't Angular's doing, it's JS's, if you don't import the service then how does JS know what Service you're trying to use?

**Comment by dschreck17** [comment]
What's the benefit of adding a service to an app.module.ts vs just importing it into a component that needs to access the service?

**Comment by nicolaisueper** [comment]
@dschreck17 Dependency Injection means that you always get the same instance of your injected class. You still have to import your files. That's part of the language.

I suggest you to research the terms 'Inversion of Control' and 'Dependency Injection'.

**Comment by gkalpak** [comment]
@dschreck17, note that the only reason youimport it into the component is because it is used as the token that tells the injector what you need. In essence, by configuring the injector (e.g. through the `NgModule`'s `providers` array) you "teach" it how to create instances of the things you need (injectables) and how to identify them. It sounds something like this (on a very high level):

1. In injector language: `providers: [{provide: 'foo', useValue: 'bar'}]`.
    In english: Dear injector, when someone (e.g. a component or other service) asks you for the injectable thing identified as `'foo'`, please pass them `'bar'`. (Note that `'foo'` and `'bar'` are strings here, but they can actuall be anything.)

2. In injector language: `providers: [{provide: 'foo', useFactory: function barFactory() { return 'bar'; }}]`.
    In english: Dear injector, when someone asks you for the injectable thing identified as `'foo'`, please execute the `barFactory` function and pass them the result. (Again, although `'foo'` is still a string it could be anything.)

3. In injector language: `providers: [{provide: 'foo', useClass: class MyFooThing {}}]`.
    In english: Dear injector, when someone asks you for the injectable thing identified as `'foo'`, please use `new MyFooThing()` to create an instance and pass them that.

4. In injector language: `providers: [{provide: MyFooThing, useClass: MyFooThing}]` (assuming you have previously defined `class MyFooThing { ... }`).
    In english: Dear injector, when someone asks you for the injectable thing identified as `MyFooThing`, please use `new MyFooThing()` to create an instance and pass them that. (Note that here the "identifier" is the class itself, not a string.)

5. Since (4) is a very common usecase, there is a shorthand. The following is equivalent to (4):
    `providers: [MyFooThing]`

One more thing to note is that each injector (in an Angular app you can have multiple injectors, but that's a different story) will create the injectable "thing" once - the first time it is requested - and then it will pass the same object to whomever asks for it. So, if two different components request the `'foo'` thing, they will get the same instance.

##
So, if you used a string as the identifier when defining the provider, you wouldn't have to inject the class into the components that want to request for an instance of that injectable. (This is for example how the AngularJS (1.x) injector works.)

This is not ideal, because it does not prevent conflicts (e.g. two 3rd-party libraries defining an injectable with the same identifier). By using the class as "identifier", you can ensure that the identifier is unique.

And finally, another reason you import the service class is to use it for type annotation, so that TypeScript knows what is the type of your argument and provide proper type-checking, intellisense etc.

##
Leaving it open for the docs team to decide whether there is room for making things more clear in the guide.

**Comment by dschreck17** [comment]
Thank you all for the explanations this has been very helpful.

**Comment by jenniferfell** [comment]
@kapunahelewong Have you already done updates related to this issue? If not, @jbogarthyde is doing some updates to the DI docs and might be able to incorporate some/all of what's requested here.

**Comment by kapunahelewong** [comment]
@jenniferfell nope, my work's on injector trees, which is in the realm of NgModules so it won't go in the DI documentation.

**Comment by kapunahelewong** [comment]
@jenniferfell This issue will need assessment to see what the impact is on the docs.

**Comment by jenniferfell** [comment]
@jbogarthyde While you're looking at DI doc, I thought this thread might be interesting/relevant. Thanks.

**Comment by kapunahelewong** [comment]
Related to: #28700 

This issue is addressed in some ways in #28700, but leaving this issue open because @gkalpakʻs explanation will be a good follow on once #28700 is merged.

**Comment by aikidave** [comment]
Closing due to inactivity and because we recently updated the dependency injection documentation.
